File: server/python/app/infrastructure/mappers/hosted_checkout_mapper.py
from onlinepayments.sdk.domain.address import Address
from onlinepayments.sdk.domain.address_personal import AddressPersonal
from onlinepayments.sdk.domain.amount_of_money import AmountOfMoney
from onlinepayments.sdk.domain.create_hosted_checkout_request import CreateHostedCheckoutRequest
from onlinepayments.sdk.domain.create_hosted_checkout_response import CreateHostedCheckoutResponse
from onlinepayments.sdk.domain.customer import Customer
from onlinepayments.sdk.domain.get_hosted_checkout_response import GetHostedCheckoutResponse
from onlinepayments.sdk.domain.hosted_checkout_specific_input import HostedCheckoutSpecificInput
from onlinepayments.sdk.domain.order import Order
from onlinepayments.sdk.domain.personal_information import PersonalInformation
from onlinepayments.sdk.domain.personal_name import PersonalName
from onlinepayments.sdk.domain.shipping import Shipping
from app.application.domain.enums.common.country import Country
from app.application.dtos.hosted_checkout.request_dto import CreateHostedCheckoutRequestDto
from app.application.dtos.hosted_checkout.response_dto import CreateHostedCheckoutResponseDto
from app.application.dtos.get_payment_by_hosted_checkout_id.response_dto import GetPaymentByHostedCheckoutIdResponseDto

class HostedCheckoutMapper:

    @staticmethod
    def to_sdk_request(request_dto: CreateHostedCheckoutRequestDto) -> CreateHostedCheckoutRequest:
        sdk_request = CreateHostedCheckoutRequest()

        order = Order()

        amount = AmountOfMoney()
        amount.amount = request_dto.amount
        if request_dto.currency is not None:
            amount.currency_code = request_dto.currency.to_currency()
        order.amount_of_money = amount

        customer = Customer()

        if request_dto.billing_address:
            personal_information = PersonalInformation()
            name = PersonalName()
            name.first_name = request_dto.billing_address.first_name
            name.surname = request_dto.billing_address.last_name
            personal_information.name = name
            customer.personal_information = personal_information

            billing_address = Address()
            billing_address.city = request_dto.billing_address.city
            if request_dto.billing_address.country is not None:
                billing_address.country_code = Country.to_iso_alpha2(request_dto.billing_address.country)
            billing_address.street = request_dto.billing_address.street
            billing_address.zip = request_dto.billing_address.zip
            customer.billing_address = billing_address

        order.customer = customer

        if request_dto.shipping_address:
            shipping = Shipping()
            shipping_address = AddressPersonal()
            shipping_address.city = request_dto.shipping_address.city
            if request_dto.shipping_address.country is not None:
                shipping_address.country_code = Country.to_iso_alpha2(request_dto.shipping_address.country)
            shipping_address.street = request_dto.shipping_address.street
            shipping_address.zip = request_dto.shipping_address.zip

            shipping_name = PersonalName()
            shipping_name.first_name = request_dto.shipping_address.first_name
            shipping_name.surname = request_dto.shipping_address.last_name
            shipping_address.name = shipping_name

            shipping.address = shipping_address
            order.shipping = shipping

        sdk_request.order = order

        hosted_checkout_specific_input = HostedCheckoutSpecificInput()
        hosted_checkout_specific_input.return_url = request_dto.redirect_url
        hosted_checkout_specific_input.locale = request_dto.language.to_locale()
        sdk_request.hosted_checkout_specific_input = hosted_checkout_specific_input

        return sdk_request

    @staticmethod
    def from_sdk_create_response(sdk_response: CreateHostedCheckoutResponse) -> CreateHostedCheckoutResponseDto:
        return CreateHostedCheckoutResponseDto(
            hosted_checkout_id=sdk_response.hosted_checkout_id,
            redirect_url=sdk_response.redirect_url,
            return_mac=sdk_response.returnmac
        )

    @staticmethod
    def from_sdk_get_response(response: GetHostedCheckoutResponse) -> GetPaymentByHostedCheckoutIdResponseDto:
        dto = GetPaymentByHostedCheckoutIdResponseDto()
        dto.status = getattr(response, 'status', None)
        dto.payment_id = getattr(response, 'created_payment_output', None) and \
                         response.created_payment_output.payment and \
                         response.created_payment_output.payment.id
        return dto

File: server/ruby/app/infrastructure/mappers/hosted_checkout_mapper.rb
module Infrastructure
  module Mappers
    module HostedCheckoutMapper
      module_function

      def to_sdk_request(dto)
        sdk_request = OnlinePayments::SDK::Domain::CreateHostedCheckoutRequest.new

        order = OnlinePayments::SDK::Domain::Order.new

        amount = OnlinePayments::SDK::Domain::AmountOfMoney.new
        amount.amount = dto.amount
        amount.currency_code = dto.currency
        order.amount_of_money = amount

        customer = OnlinePayments::SDK::Domain::Customer.new

        if dto.billing_address
          personal = OnlinePayments::SDK::Domain::PersonalInformation.new
          name = OnlinePayments::SDK::Domain::PersonalName.new
          name.first_name = dto.billing_address&.first_name
          name.surname = dto.billing_address&.last_name
          personal.name = name
          customer.personal_information = personal

          billing = OnlinePayments::SDK::Domain::Address.new
          billing.city = dto.billing_address&.city
          billing.country_code = Business::Extensions::Models::CountryExtension.to_iso_alpha2(dto.billing_address&.country)
          billing.street = dto.billing_address&.street
          billing.zip = dto.billing_address&.zip
          customer.billing_address = billing
        end

        order.customer = customer

        if dto.shipping_address
          shipping = OnlinePayments::SDK::Domain::Shipping.new
          shipping_address = OnlinePayments::SDK::Domain::AddressPersonal.new
          shipping_address.city = dto.shipping_address&.city
          shipping_address.country_code = Business::Extensions::Models::CountryExtension.to_iso_alpha2(dto.shipping_address&.country)
          shipping_address.street = dto.shipping_address&.street
          shipping_address.zip = dto.shipping_address&.zip

          shipping_name = OnlinePayments::SDK::Domain::PersonalName.new
          shipping_name.first_name = dto.shipping_address&.first_name
          shipping_name.surname = dto.shipping_address&.last_name
          shipping_address.name = shipping_name

          shipping.address = shipping_address
          order.shipping = shipping
        end

        hosted_checkout_input = OnlinePayments::SDK::Domain::HostedCheckoutSpecificInput.new
        hosted_checkout_input.return_url = dto.redirect_url
        hosted_checkout_input.locale = Business::Extensions::Models::LanguageExtension.to_locale(dto.language, dto.billing_address&.country)

        sdk_request.order = order
        sdk_request.hosted_checkout_specific_input = hosted_checkout_input

        sdk_request
      end

      def from_sdk_create_response(response)
        Business::Dtos::CreateHostedCheckout::ResponseDto.new(
          hosted_checkout_id: response.hosted_checkout_id,
          redirect_url: response.redirect_url,
          return_mac: response.returnmac
        )
      end

      def from_sdk_get_response(response)
        dto = Business::Dtos::GetPaymentByHostedCheckoutId::ResponseDto.new
        dto.status = response&.status
        dto.payment_id = response&.created_payment_output&.payment&.id
        dto
      end
    end
  end
end